Document the 'experimental.connpool' module (#4329)
tarantool/doc: doc/reference/reference_lua/connpool.rst

..  _connpool_module:

Module experimental.connpool
============================

**Since:** :doc:`3.1.0 </release/3.1.0>`

..  important::

    ``experimental.connpool`` is an experimental module and is subject to changes.

> An ``experimental.connpool`` module provides a set of features for connecting to remote cluster instances and for executing remote procedure calls on an instance that satisfies the specified criteria.

Reviewer: ```suggestion
The ``experimental.connpool`` module provides a set of features for connecting to remote cluster instances and for executing remote procedure calls on an instance that satisfies the specified criteria.
```

Reviewer: I see that elsewhere in the doc we keep using a definite article before module names, even when introduced for the first time (e.g. https://docs.d.tarantool.io/en/doc/connpool-api/reference/reference_lua/iconv/), with the only exception of explicitly adding "new" like in "a new XXX module" in release notes, but not in the main doc.

..  _connpool_module_load:

Loading a module
----------------

To load the ``experimental.connpool`` module, use the ``require()`` directive:

..  code-block:: lua

    local connpool = require('experimental.connpool')

..  _connpool_module_api_reference:

API Reference
-------------

..  container:: table

    ..  rst-class:: left-align-column-1
    ..  rst-class:: left-align-column-2

    ..  list-table::
        :widths: 35 65

        *   -   **Functions**
            -

        *   -   :ref:`connpool.call() <connpool_module_call>`
            -   Execute the specified function on a remote instance

        *   -   :ref:`connpool.connect() <connpool_module_connect>`
            -   Create a connection to the specified instance

        *   -   :ref:`connpool.filter() <connpool_module_filter>`
            -   Get names of instances that match the specified conditions

.. module:: connpool

..  _connpool_module_api_reference_functions:

Functions
~~~~~~~~~

..  _connpool_module_call:

..  function:: call(func_name, args, opts)

    Execute the specified function on a remote instance.

    ..  NOTE::

>         You need to grant the ``execute`` :ref:`permission <configuration_credentials_managing_users_roles_granting_privileges>` for the specified function to a user used for replication to execute this function on a remote instance.

Reviewer: ```suggestion
        You need to grant the ``execute`` :ref:`permission <configuration_credentials_managing_users_roles_granting_privileges>` for the specified function to users with the ``replication`` :ref:`role <box_space-user>` to execute this function on a remote instance.
```

Author: Technically speaking, a _user used for replication_ and a _user with the replication role_ might be different entities. In particular, _user used for replication_ maintains inter-cluster communications while a _user with the replication role_ can be created to allow some third-party system to get data from a Tarantool cluster (for example, CDC).

                  labels = { dc = 'east' },
                  mode = 'ro' }
        )

..  _connpool_module_connect:

..  function:: connect(instance_name, opts)

    Create a connection to the specified instance.

    :param string instance_name: an instance name.
    :param table/nil opts: none, any, or all of the following parameters:

                           * ``connect_timeout`` -- a connection timeout (in seconds).
                           * ``wait_connected`` -- whether to block the connection until it is established:

                               * if ``true`` (default), the connection is blocked until it is established.
                               * if ``false``, the connection is returned immediately.

                           * ``fetch_schema`` -- whether to fetch schema changes from a remote instance.

    :return: a :ref:`net.box <net_box-module>` connection.

    **Example**

    In the example below, ``connect()`` is used to create the active connection to ``storage-b-002``:

    ..  code-block:: lua

        local connpool = require('experimental.connpool')
        local conn = connpool.connect("storage-b-002", { fetch_schema = true })

>     Once you have a connection, you can execute requests on a remote instance, for example, select data from a space using :ref:`conn.space.\<space-name\>:select() <conn-select>`.

Reviewer: ```suggestion
    Once you have a connection, you can execute requests on the remote instance, for example, select data from a space using :ref:`conn.space.\<space-name\>:select() <conn-select>`.
```

Reviewer: Meaning "the instance to which the connection is established". As I understand, within a single `connect()` function call a connection can be established to only 1 instance, so the instance is unique in this context.
